### ISSUE TITLE
ONNX failures in cadene models

### ISSUE THREAD

**Issue by mvermeulen** [ROCmSoftwarePlatform/AMDMIGraphX#119]
This issue documents the models that fail "read_onnx" when ONNX files are exported from the cadene pretrained repository (these might be known issues, so can resolve as duplicate when fixed, but leave as documentation until they are fixed):

nasnetalarge: operators.hpp:334: Non-axis dimensions should match
nasnetamobile: operators.hpp:334: Non-axis dimensions should match
pnasnet5large: operators.hpp:334: Non-axis dimensions should match

Symptoms and naming of the files suggest these are likely the same issue.

This issue also documents the models that fail "perf_onnx" at runtime when ONNX files are exercised:

dpn107, dpn131, dpn68b, dpn92, dpn98, resnext101_32x4d, resnext101_64x4d -
Memory access fault by GPU node-2 (Agent handle: 0x1882a50) on address 0x5048ea0
00. Reason: Page not present or supervisor privilege.

fbresnet152 -
terminate called after throwing an instance of 'migraphx::version_1::exception'
  what():  /data/repo/AMDMIGraphX/src/program.cpp:310: Invalid program from comp
ilation at instruction 1028

inceptionresnetv2 -
terminate called after throwing an instance of 'migraphx::version_1::exception'
  what():  /data/repo/AMDMIGraphX/src/program.cpp:310: Invalid program from comp
ilation at instruction 2020

Separate from that there are a few models that fail to export due to a pytorch/onnx issue unrelated to migraphx (https://github.com/pytorch/pytorch/issues/2898).  These include: bninception, squeezenet, xception, senet, se_resnet, se_resnext, cafferesnet.  Included here for completeness.

**Comment by mvermeulen** [comment]
It looks like dpn107, resnext101_64x4d now pass.  I didn't test all the dpn* resnext* models, but assume these now pass.

Also some additional diagnosis for fbresnet152 and inceptionresnetv2 -
1. Both will pass if "common_subexpression_elimination{}" is commented out in src/targets/gpu/target.cpp
2. Run with "env MIGRAPHX_TRACE_COMPILE=1" and the program fails during the common_subexpression_elimination pass when printing the instruction stream.  The last instructions printed are:

@974 = relu(@973) -> float_type, {1, 256, 56, 56}, {802816, 3136, 56, 1}
@975 = convolution[padding=0, 0,stride=1, 1,dilation=1, 1,padding_mode=0,group=1](@974,@883) -> float_type, {1, 64, 56, 56}, {200704, 3136, 56, 1}
@976 = broadcast[axis=1](@876) -> float_type, {1, 64, 56, 56}, {0, 1, 0, 0}
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
@977 = add(@975,Aborted (core dumped)

Is CSE properly handling these broadcast nodes? Is something else going on during CSE?

**Comment by mvermeulen** [comment]
dpn107 no longer works.  I see the following error reported:

terminate called after throwing an instance of 'migraphx::version_1::exception'
  what():  /home/mev/source/AMDMIGraphX/src/include/migraphx/operators.hpp:895: when broadcasting success sizes must match

**Comment by pfultz2** [comment]
> Were all the necessary changes included in pull request 150? Was the merge done correctly?

We cant merge to develop unless its updated to the latest which is why the previous commit merges develop into the PR first.

**Comment by kahmed10** [comment]
The broadcast issue seems to be affecting all dpn models with batch size > 1. I will look into this.

**Comment by mvermeulen** [comment]
Latest status based on 90a37082deb7c2e2a410545352ef344820c67e64:

fbresnet152 -
terminate called after throwing an instance of 'migraphx::version_1::exception'
what(): /data/repo/AMDMIGraphX/src/program.cpp:310: Invalid program from comp
ilation at instruction 1028

inceptionresnetv2 -
terminate called after throwing an instance of 'migraphx::version_1::exception'
what(): /data/repo/AMDMIGraphX/src/program.cpp:310: Invalid program from comp
ilation at instruction 2020

All others including
-- nasnetalarge, nasnetamobile, pnasnet5large 
-- dpn107, dpn131, dpn92, dpn98 ... 
Now pass perf_onnx

**Comment by mvermeulen** [comment]
Spoke too soon...  dpn92 does not at least for largest batch sizes whereas dpn98, dpn107, dpn131, dpn68 do work.  This could be side effect of having a tuning run.
```
mev@chilecito:~/source/AMDMIGraphX/build/src/onnx$ ./perf_onnx /home/mev/source/migraphx_onnx/cadene/dpn92i16.onnx 
[libprotobuf INFO google/protobuf/io/coded_stream.cc:610] Reading dangerously large protocol message.  If the message turns out to be larger than 2147483647 bytes, parsing will be halted for security reasons.  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:81] The total number of bytes read was 152000513
Compiling ... 
Allocating params ... 
Running performance report ... 
Memory access fault by GPU node-2 (Agent handle: 0x20fa950) on address 0x7f6e2d41f000. Reason: Page not present or supervisor privilege.
Aborted (core dumped)
mev@chilecito:~/source/AMDMIGraphX/build/src/onnx$ ./perf_onnx /home/mev/source/migraphx_onnx/cadene/dpn107i16.onnx 
[libprotobuf INFO google/protobuf/io/coded_stream.cc:610] Reading dangerously large protocol message.  If the message turns out to be larger than 2147483647 bytes, parsing will be halted for security reasons.  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:81] The total number of bytes read was 349482025
Compiling ... 
Allocating params ... 
Running performance report ... 
@0 = check_context -> float_type, {}, {}: 0.000693077ms, 1%
scratch = @param:scratch -> float_type, {69443584}, {1}: 0.00114385ms, 1%
@2 = hip::load_literal -> float_type, {1600}, {1}: 0.000505769ms, 1%
...
```
dpn97 batch size 1 seems to work
dpn97 batch size 2 seems to work
dpn97 batch size 4 seems to work
dpn97 batch size 8 fails
dpn97 batch size 16 fails
dpn97 batch size 32 fails
dpn107 batch size 16 seems to work
dpn98 batch size 16 seems to work
dpn131 batch size 16 seems to work
dpn16 batch size 16 seems to work

One potential difference between dpn97 and other dpn networks is that I ran a MIOPEN_FIND_ENFORCE step to set up tuned kernels (it might have worked before that, don't 100% recall what I ran before tuning).

**Comment by mvermeulen** [comment]
Cadene models now all pass (at least the ones I can export).  Closing this issue.
